petastorm/reader_impl/shuffling_buffer.py

```diff
@@ -0,0 +1,168 @@
+#  Copyright (c) 2017-2018 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+import numpy as np
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ShufflingBufferBase(object):
+    """Shuffling implements a shuffling algorithm. Items can be added to the shuffling buffer and removed in a
+    different order as defined by the concrete shuffling algorithm."""
+    @abc.abstractmethod
+    def add_many(self, items):
+        """Adds multiple items to the buffer.
+
+        :param items: items to be added to the shuffling buffer.
+        :return: None
+        """
+        pass
+
+    @abc.abstractmethod
+    def retrieve(self):
+        """Selects an item from the buffer and returns the item to the caller. The item is removed from the buffer.
+
+        :return: The selected item.
+        """
+        pass
+
+    @abc.abstractmethod
+    def can_add(self):
+        """Checks the state of the buffer and returns whether a new item can be added to the buffer at the time.
+
+        :return: A boolean indicating whether an item can be added to the buffer at the time.
+        """
+        pass
+
+    @abc.abstractmethod
+    def can_retrieve(self):
+        """Checks the state of the buffer and returns whether an item can be removed from the buffer..
+
+        :return: A boolean indicating whether an item can be returned from the buffer at the time.
+        """
+        pass
+
+    @abc.abstractproperty
+    def size(self):
+        """Returns the number of elements currently present in the buffer.
+
+        :return: number of elements currently present in the buffer
+        """
+        pass
+
+    @abc.abstractmethod
+    def finish(self):
+        """Call this method when no more `enqueu_many` calls will be made.
```

Comment:
Added a clarification comment:
> This allows a user to deplete the buffer. Typically during last epoch. Otherwise, we would always have leftovers in the buffer at the end of the lifecycle.